File: weblayer/browser/java/org/chromium/weblayer_private/FullscreenDelegateProxy.java
// Copyright 2019 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.weblayer_private;

import android.os.RemoteException;
import android.webkit.ValueCallback;

import org.chromium.base.annotations.CalledByNative;
import org.chromium.base.annotations.JNINamespace;
import org.chromium.base.annotations.NativeMethods;
import org.chromium.weblayer_private.aidl.IFullscreenDelegateClient;
import org.chromium.weblayer_private.aidl.ObjectWrapper;

/**
 * Owns the c++ FullscreenDelegateProxy class, which is responsible for forwarding all
 * FullscreenDelegate delegate calls to this class, which in turn forwards to the
 * FullscreenDelegateClient.
 */
@JNINamespace("weblayer")
public final class FullscreenDelegateProxy {
    private long mNativeFullscreenDelegateProxy;
    private IFullscreenDelegateClient mClient;

    FullscreenDelegateProxy(long browserController, IFullscreenDelegateClient client) {
        assert client != null;
        mClient = client;
        mNativeFullscreenDelegateProxy =
                FullscreenDelegateProxyJni.get().createFullscreenDelegateProxy(
                        this, browserController);
    }

    public void setClient(IFullscreenDelegateClient client) {
        assert client != null;
        mClient = client;
    }

    public void destroy() {
        FullscreenDelegateProxyJni.get().deleteFullscreenDelegateProxy(
                mNativeFullscreenDelegateProxy);
        mNativeFullscreenDelegateProxy = 0;
    }

    @CalledByNative
    private void enterFullscreen() throws RemoteException {
        ValueCallback<Void> exitFullscreenCallback = new ValueCallback<Void>() {
            @Override
            public void onReceiveValue(Void result) {
                if (mNativeFullscreenDelegateProxy == 0) {
                    throw new IllegalStateException("Called after destroy()");
                }
                FullscreenDelegateProxyJni.get().doExitFullscreen(
                        mNativeFullscreenDelegateProxy, FullscreenDelegateProxy.this);
            }
        };
        mClient.enterFullscreen(ObjectWrapper.wrap(exitFullscreenCallback));
    }

    @CalledByNative
    private void exitFullscreen() throws RemoteException {
        mClient.exitFullscreen();
    }

    @NativeMethods
    interface Natives {
        long createFullscreenDelegateProxy(FullscreenDelegateProxy proxy, long browserController);
        void deleteFullscreenDelegateProxy(long proxy);
        void doExitFullscreen(long nativeFullscreenDelegateProxy, FullscreenDelegateProxy proxy);
    }
}

File: weblayer/browser/fullscreen_delegate_proxy.cc
// Copyright 2019 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "weblayer/browser/fullscreen_delegate_proxy.h"

#include "base/android/jni_string.h"
#include "url/gurl.h"
#include "weblayer/browser/browser_controller_impl.h"
#include "weblayer/browser/java/jni/FullscreenDelegateProxy_jni.h"

using base::android::AttachCurrentThread;
using base::android::ConvertUTF8ToJavaString;
using base::android::ScopedJavaLocalRef;

namespace weblayer {

FullscreenDelegateProxy::FullscreenDelegateProxy(
    JNIEnv* env,
    jobject obj,
    BrowserController* browser_controller)
    : browser_controller_(browser_controller), java_observer_(env, obj) {
  browser_controller_->SetFullscreenDelegate(this);
}

FullscreenDelegateProxy::~FullscreenDelegateProxy() {
  browser_controller_->SetFullscreenDelegate(nullptr);
}

void FullscreenDelegateProxy::EnterFullscreen(base::OnceClosure exit_closure) {
  exit_fullscreen_closure_ = std::move(exit_closure);
  Java_FullscreenDelegateProxy_enterFullscreen(AttachCurrentThread(),
                                               java_observer_);
}

void FullscreenDelegateProxy::ExitFullscreen() {
  Java_FullscreenDelegateProxy_exitFullscreen(AttachCurrentThread(),
                                              java_observer_);
}

void FullscreenDelegateProxy::DoExitFullscreen(
    JNIEnv* env,
    const base::android::JavaParamRef<jobject>& caller) {
  if (exit_fullscreen_closure_)
    std::move(exit_fullscreen_closure_).Run();
}

static jlong JNI_FullscreenDelegateProxy_CreateFullscreenDelegateProxy(
    JNIEnv* env,
    const base::android::JavaParamRef<jobject>& proxy,
    jlong browser_controller) {
  return reinterpret_cast<jlong>(new FullscreenDelegateProxy(
      env, proxy,
      reinterpret_cast<BrowserControllerImpl*>(browser_controller)));
}

static void JNI_FullscreenDelegateProxy_DeleteFullscreenDelegateProxy(
    JNIEnv* env,
    jlong proxy) {
  delete reinterpret_cast<FullscreenDelegateProxy*>(proxy);
}

}  // namespace weblayer
